packages/lib/server/getLuckyUser.ts

```diff
@@ -114,18 +103,84 @@ function getUsersWithHighestPriority<T extends Pick<User, "id" | "email"> & { pr
+
+  // Calculate the total weightAdjustments of all round-robin hosts
+  const allWeightAdjustments = allRRHosts.reduce((sum, host) => sum + (host.weightAdjustment ?? 0), 0);
+
+  // Calculate the total weight of all round-robin hosts
+  const totalWeight = allRRHosts.reduce((sum, host) => sum + (host.weight ?? 100), 0);
+
+  // Calculate booking shortfall for each available user
+  const usersWithBookingShortfalls = availableUsers.map((user) => {
+    const targetPercentage = (user.weight ?? 100) / totalWeight;
+
+    const userBookings = bookings.filter(
+      (booking) =>
+        booking.userId === user.id || booking.attendees.some((attendee) => attendee.email === user.email)
+    );
+
+    const targetNumberOfBookings = (bookings.length + allWeightAdjustments) * targetPercentage;
+    const bookingShortfall = targetNumberOfBookings - (userBookings.length + (user.weightAdjustment ?? 0));
+
+    return {
+      ...user,
+      bookingShortfall,
+    };
+  });
+
+  // Find users with the highest booking shortfall
+  const maxShortfall = Math.max(...usersWithBookingShortfalls.map((user) => user.bookingShortfall));
+  const userIdsWithMaxShortfall = new Set(
+    usersWithBookingShortfalls.filter((user) => user.bookingShortfall === maxShortfall).map((user) => user.id)
+  );
+
+  return availableUsers.filter((user) => userIdsWithMaxShortfall.has(user.id));
+}
+
 // TODO: Configure distributionAlgorithm from the event type configuration
 // TODO: Add 'MAXIMIZE_FAIRNESS' algorithm.
-export async function getLuckyUser<T extends Pick<User, "id" | "email"> & { priority?: number | null }>(
+export async function getLuckyUser<
+  T extends PartialUser & {
+    priority?: number | null;
+    weight?: number | null;
+    weightAdjustment?: number | null;
+  }
+>(
   distributionAlgorithm: "MAXIMIZE_AVAILABILITY" = "MAXIMIZE_AVAILABILITY",
-  { availableUsers, eventTypeId }: { availableUsers: T[]; eventTypeId: number }
+  getLuckyUserParams: GetLuckyUserParams<T>
 ) {
+  const { availableUsers, eventType, allRRHosts } = getLuckyUserParams;
+
   if (availableUsers.length === 1) {
     return availableUsers[0];
   }
+
+  // all bookings of event type of all rr hosts
+  const allBookings = await BookingRepository.getAllBookingsOfUsers({
+    eventTypeId: eventType.id,
+    users: allRRHosts.map((host) => {
+      return { id: host.user.id, email: host.user.email };
+    }),
+  });
+
   switch (distributionAlgorithm) {
     case "MAXIMIZE_AVAILABILITY":
-      const highestPriorityUsers = getUsersWithHighestPriority({ availableUsers });
-      return leastRecentlyBookedUser<T>({ availableUsers: highestPriorityUsers, eventTypeId });
+      let possibleLuckyUsers = availableUsers;
+      if (eventType.isRRWeightsEnabled) {
+        possibleLuckyUsers = await getUsersBasedOnWeights({ ...getLuckyUserParams, allBookings });
+      }
+      const highestPriorityUsers = getUsersWithHighestPriority({ availableUsers: possibleLuckyUsers });
```

Comment:
if `getUsersBasedOnWeights` returned more users, pick the one with highest priority